File: webflux-dialogue/src/main/java/com/study/webflux/rag/application/dialogue/pipeline/stage/SystemPromptService.java
package com.study.webflux.rag.application.dialogue.pipeline.stage;

import java.util.List;
import java.util.StringJoiner;

import lombok.extern.slf4j.Slf4j;

import org.springframework.stereotype.Service;

import com.study.webflux.rag.application.dialogue.pipeline.SystemPromptContext;
import com.study.webflux.rag.application.dialogue.policy.PromptTemplatePolicy;
import com.study.webflux.rag.domain.dialogue.model.PersonaId;
import com.study.webflux.rag.domain.dialogue.port.TemplateLoaderPort;
import com.study.webflux.rag.domain.memory.model.Memory;
import com.study.webflux.rag.domain.memory.model.MemoryRetrievalResult;
import com.study.webflux.rag.domain.retrieval.model.RetrievalContext;

@Slf4j
@Service
public class SystemPromptService {

	private static final String PERSONA_TEMPLATE_PREFIX = "system/persona/";
	private static final String PERSONA_PLACEHOLDER = "{{persona}}";
	private static final String COMMON_PLACEHOLDER = "{{common}}";
	private static final String MEMORIES_PLACEHOLDER = "{{memories}}";
	private static final String CONTEXT_PLACEHOLDER = "{{context}}";

	private static final String MEMORIES_TITLE = "대화 상대에 대한 기억:";
	private static final String EXPERIENTIAL_MEMORY_TITLE = "경험적 기억:";
	private static final String FACTUAL_MEMORY_TITLE = "사실 기반 기억:";
	private static final String CONTEXT_TITLE = "참고 정보:";

	private final TemplateLoaderPort templateLoader;
	private final PromptTemplatePolicy policy;

	public SystemPromptService(TemplateLoaderPort templateLoader, PromptTemplatePolicy policy) {
		this.templateLoader = templateLoader;
		this.policy = policy;
	}

	/**
	 * 페르소나별 시스템 프롬프트를 생성합니다.
	 *
	 * @param personaId
	 *            페르소나 ID (템플릿 파일명으로 사용)
	 * @param context
	 *            현재 질의에 대한 검색 컨텍스트
	 * @param memories
	 *            현재 대화 상대에 대한 회수 메모리
	 * @return LLM 요청에 주입할 시스템 프롬프트
	 */
	public String buildSystemPrompt(SystemPromptContext context) {
		PromptSections sections = new PromptSections(resolvePersonaPrompt(context.personaId()),
			policy.commonTemplate(),
			buildMemoryBlock(context.memoryResult()),
			buildContextBlock(context.retrievalContext()));

		String baseTemplate = policy.baseTemplate();
		if (!baseTemplate.isBlank()) {
			return applyTemplate(baseTemplate, sections);
		}

		return joinNonBlankBlocks(sections.persona(),
			sections.common(),
			sections.memories(),
			sections.context());
	}

	/**
	 * 기본 페르소나로 시스템 프롬프트를 생성합니다. (하위 호환성)
	 */
	private String resolvePersonaPrompt(PersonaId personaId) {
		if (personaId != null && !PersonaId.DEFAULT.equals(personaId)) {
			String dynamicTemplate = loadPersonaTemplate(personaId);
			if (!dynamicTemplate.isBlank()) {
				return dynamicTemplate;
			}
		}
		return resolveDefaultPersonaPrompt();
	}

	/**
	 * 기본 페르소나 프롬프트를 템플릿 우선 정책으로 선택합니다.
	 */
	private String resolveDefaultPersonaPrompt() {
		String defaultPersonaTemplate = policy.defaultPersonaTemplate();
		if (!defaultPersonaTemplate.isBlank()) {
			return defaultPersonaTemplate;
		}
		String configuredSystemPrompt = policy.configuredSystemPrompt();
		if (!configuredSystemPrompt.isBlank()) {
			return configuredSystemPrompt.trim();
		}
		return "";
	}

	/**
	 * personaId에 해당하는 템플릿 파일을 로드합니다.
	 */
	private String loadPersonaTemplate(PersonaId personaId) {
		String templatePath = PERSONA_TEMPLATE_PREFIX + personaId.value();
		try {
			return templateLoader.load(templatePath).trim();
		} catch (RuntimeException e) {
			log.warn("페르소나 템플릿 '{}'을 불러오지 못해 기본 페르소나를 사용합니다: {}",
				templatePath,
				e.getMessage());
			return "";
		}
	}

	/**
	 * 메모리 조회 결과를 시스템 프롬프트 블록 문자열로 직렬화합니다.
	 */
	private String buildMemoryBlock(MemoryRetrievalResult memories) {
		if (memories == null || memories.isEmpty()) {
			return "";
		}

		StringBuilder builder = new StringBuilder(MEMORIES_TITLE).append("\n");
		appendMemorySection(builder, EXPERIENTIAL_MEMORY_TITLE, memories.experientialMemories());
		appendMemorySection(builder, FACTUAL_MEMORY_TITLE, memories.factualMemories());

		return builder.toString().trim();
	}

	/**
	 * 검색 결과 문서 내용을 시스템 프롬프트 블록 문자열로 직렬화합니다.
	 */
	private String buildContextBlock(RetrievalContext context) {
		if (context == null || context.isEmpty()) {
			return "";
		}
		StringJoiner joiner = new StringJoiner("\n");
		context.documents().stream()
			.map(doc -> doc.content())
			.filter(content -> content != null && !content.isBlank())
			.forEach(joiner::add);
		if (joiner.length() == 0) {
			return "";
		}
		return CONTEXT_TITLE + "\n" + joiner;
	}

	/**
	 * 베이스 템플릿에 각 블록을 치환하고 공백 라인을 정규화합니다.
	 */
	private String applyTemplate(String template,
		PromptSections sections) {
		String rendered = template.replace(PERSONA_PLACEHOLDER, normalizeBlock(sections.persona()))
			.replace(COMMON_PLACEHOLDER, normalizeBlock(sections.common()))
			.replace(MEMORIES_PLACEHOLDER, normalizeBlock(sections.memories()))
			.replace(CONTEXT_PLACEHOLDER, normalizeBlock(sections.context()));
		return normalizeTemplateSpacing(rendered);
	}

	/**
	 * 템플릿이 없을 때 비어 있지 않은 블록만 순서대로 이어붙입니다.
	 */
	private String joinNonBlankBlocks(String... blocks) {
		StringJoiner joiner = new StringJoiner("\n\n");
		for (String block : blocks) {
			if (block != null && !block.isBlank()) {
				joiner.add(block.trim());
			}
		}
		return joiner.toString();
	}

	private void appendMemorySection(StringBuilder builder,
		String sectionTitle,
		List<Memory> memories) {
		if (memories == null || memories.isEmpty()) {
			return;
		}
		List<String> lines = memories.stream()
			.map(Memory::content)
			.filter(content -> content != null && !content.isBlank())
			.map(content -> "- " + content)
			.toList();
		if (lines.isEmpty()) {
			return;
		}
		builder.append("\n").append(sectionTitle).append("\n");
		lines.forEach(line -> builder.append(line).append("\n"));
	}

	private String normalizeBlock(String block) {
		return block == null ? "" : block.trim();
	}

	private String normalizeTemplateSpacing(String template) {
		String compacted = template.replaceAll("(?m)^[ \\t]+$", "");
		compacted = compacted.replaceAll("\\n{3,}", "\n\n");
		return compacted.trim();
	}

	private record PromptSections(
		String persona,
		String common,
		String memories,
		String context) {
	}
}

File: app-miyou/src/main/kotlin/com/miyou/app/application/dialogue/pipeline/stage/SystemPromptService.kt
package com.miyou.app.application.dialogue.pipeline.stage

import com.miyou.app.application.dialogue.policy.PromptTemplatePolicy
import com.miyou.app.domain.dialogue.model.PersonaId
import com.miyou.app.domain.dialogue.port.TemplateLoaderPort
import com.miyou.app.domain.memory.model.Memory
import com.miyou.app.domain.memory.model.MemoryRetrievalResult
import com.miyou.app.domain.retrieval.model.RetrievalContext
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.util.StringJoiner

@Service
class SystemPromptService(
    private val templateLoader: TemplateLoaderPort,
    private val policy: PromptTemplatePolicy,
) {
    fun buildSystemPrompt(
        personaId: PersonaId,
        context: RetrievalContext,
        memories: MemoryRetrievalResult,
    ): String {
        val sections =
            PromptSections(
                resolvePersonaPrompt(personaId),
                policy.commonTemplate,
                buildMemoryBlock(memories),
                buildContextBlock(context),
            )

        val baseTemplate = policy.baseTemplate
        if (!baseTemplate.isBlank()) {
            return applyTemplate(baseTemplate, sections)
        }
        return joinNonBlankBlocks(sections.persona, sections.common, sections.memories, sections.context)
    }

    fun buildSystemPrompt(
        context: RetrievalContext,
        memories: MemoryRetrievalResult,
    ): String = buildSystemPrompt(PersonaId.defaultPersona(), context, memories)

    private fun resolvePersonaPrompt(personaId: PersonaId): String {
        if (personaId.value != PersonaId.DEFAULT.value) {
            val dynamicTemplate = loadPersonaTemplate(personaId)
            if (dynamicTemplate.isNotBlank()) {
                return dynamicTemplate
            }
        }
        return resolveDefaultPersonaPrompt()
    }

    private fun resolveDefaultPersonaPrompt(): String {
        val defaultPersonaTemplate = policy.defaultPersonaTemplate
        if (!defaultPersonaTemplate.isBlank()) {
            return defaultPersonaTemplate
        }
        val configuredSystemPrompt = policy.configuredSystemPrompt
        if (configuredSystemPrompt.isNotBlank()) {
            return configuredSystemPrompt.trim()
        }
        return ""
    }

    private fun loadPersonaTemplate(personaId: PersonaId): String {
        val templatePath = "$PERSONA_TEMPLATE_PREFIX${personaId.value}"
        return try {
            templateLoader.load(templatePath).trim()
        } catch (exception: RuntimeException) {
            logger.warn("Persona template load failed: {} ({})", templatePath, exception.message)
            ""
        }
    }

    private fun buildMemoryBlock(memories: MemoryRetrievalResult): String {
        if (memories.isEmpty()) {
            return ""
        }

        val builder = StringBuilder(MEMORIES_TITLE).append('\n')
        appendMemorySection(builder, EXPERIENTIAL_MEMORY_TITLE, memories.experientialMemories)
        appendMemorySection(builder, FACTUAL_MEMORY_TITLE, memories.factualMemories)
        return builder.toString().trim()
    }

    private fun buildContextBlock(context: RetrievalContext): String {
        if (context.isEmpty()) {
            return ""
        }
        val joiner = StringJoiner("\n")
        context.documents
            .map { it.content }
            .filter { !it.isNullOrBlank() }
            .forEach(joiner::add)
        if (joiner.length() == 0) {
            return ""
        }
        return "$CONTEXT_TITLE\n$joiner"
    }

    private fun applyTemplate(
        template: String,
        sections: PromptSections,
    ): String {
        val rendered =
            template
                .replace(PERSONA_PLACEHOLDER, normalizeBlock(sections.persona))
                .replace(COMMON_PLACEHOLDER, normalizeBlock(sections.common))
                .replace(MEMORIES_PLACEHOLDER, normalizeBlock(sections.memories))
                .replace(CONTEXT_PLACEHOLDER, normalizeBlock(sections.context))
        return normalizeTemplateSpacing(rendered)
    }

    private fun joinNonBlankBlocks(vararg blocks: String): String {
        val joiner = StringJoiner("\n\n")
        blocks.forEach { block ->
            if (block.isNotBlank()) {
                joiner.add(block.trim())
            }
        }
        return joiner.toString()
    }

    private fun appendMemorySection(
        builder: StringBuilder,
        sectionTitle: String,
        memories: List<Memory>,
    ) {
        if (memories.isEmpty()) {
            return
        }
        val lines =
            memories
                .mapNotNull { it.content }
                .filter { it.isNotBlank() }
                .map { "- $it" }
                .toList()
        if (lines.isEmpty()) {
            return
        }
        builder.append('\n').append(sectionTitle).append('\n')
        lines.forEach { line ->
            builder.append(line).append('\n')
        }
    }

    private fun normalizeBlock(block: String): String = block.trim()

    private fun normalizeTemplateSpacing(template: String): String {
        var compacted = template.replace(Regex("(?m)^[ \\t]+$"), "")
        compacted = compacted.replace(Regex("\\n{3,}"), "\n\n")
        return compacted.trim()
    }

    private fun normalizeTemplateInput(template: String?): String = template ?: ""

    private data class PromptSections(
        val persona: String,
        val common: String,
        val memories: String,
        val context: String,
    )

    private companion object {
        const val PERSONA_TEMPLATE_PREFIX = "system/persona/"
        const val PERSONA_PLACEHOLDER = "{{persona}}"
        const val COMMON_PLACEHOLDER = "{{common}}"
        const val MEMORIES_PLACEHOLDER = "{{memories}}"
        const val CONTEXT_PLACEHOLDER = "{{context}}"

        const val MEMORIES_TITLE = "기억 데이터:"
        const val EXPERIENTIAL_MEMORY_TITLE = "체험 기억:"
        const val FACTUAL_MEMORY_TITLE = "사실 기억:"
        const val CONTEXT_TITLE = "지금 상황:"

        private val logger by lazy { LoggerFactory.getLogger(SystemPromptService::class.java) }
    }
}
